exonum/src/node/state.rs

```diff
@@ -851,6 +909,35 @@ impl State {
         }
     }
 
+    /// Finds unknown transactions in the block.
+    pub fn create_incomplete_block(
+        &mut self,
+        msg: &BlockResponse,
+        txs: &MapIndex<&&Snapshot, Hash, RawMessage>,
+        txs_pool: &KeySetIndex<&&Snapshot, Hash>,
+    ) -> Result<bool, failure::Error> {
+        let mut unknown_txs = HashSet::new();
+        for hash in msg.transactions() {
+            if txs.get(hash).is_some() {
+                if !txs_pool.contains(hash) {
+                    panic!(
+                        "Received block with already \
+                         committed transaction"
+                    )
+                }
+            } else {
+                unknown_txs.insert(*hash);
+            }
+        }
+
+        self.incomplete_block = Some(IncompleteBlock {
+            msg: msg.clone(),
+            unknown_txs: unknown_txs.clone(),
```

Comment:
As I can see, we are cloning `unknown_txs` here just to check that it's not empty below. It can easily be avoided, e.g.:
```
let response = Ok(!unknown_txs.is_empty());
self.incomplete_block = Some(IncompleteBlock {
    msg: msg.clone(),
    unknown_txs: unknown_txs,
});
response
```